File: CSharp/Inventory_and_Order_Processing/Services/BudgetFinder.cs
using Models;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Services
{
    public class ClosestPairResult
    {
        public Product ProductLeft { get; set; }
        public Product ProductRight { get; set; }
        public double CombinedPrice { get; set; }
        public double Difference { get; set; }
    }
    public static class BudgetFinder
    {
        public static ClosestPairResult FindClosestPairToBudget(List<Product> products, double budget)
        {
            if (products.Count < 2)
            {
                Console.WriteLine("Not enough products to form a pair.");
                return null;
            }

            List<Product> sortedProducts = SortHelper.SortProductsByPrice(products);

            int left = 0;
            int right = sortedProducts.Count - 1;

            ClosestPairResult bestPair = null;
            double bestDifference = double.PositiveInfinity;

            while (left < right)
            {
                Product productLeft = sortedProducts[left];
                Product productRight = sortedProducts[right];
                double combinedPrice = productLeft.Price + productRight.Price;
                double difference = Math.Abs(combinedPrice - budget);

                if (difference < bestDifference)
                {
                    bestDifference = difference;
                    bestPair = new ClosestPairResult
                    {
                        ProductLeft = productLeft,
                        ProductRight = productRight,
                        CombinedPrice = combinedPrice,
                        Difference = difference
                    };
                }

                if (combinedPrice < budget)
                {
                    left++;
                }
                else if (combinedPrice > budget)
                {
                    right--;
                }
                else
                {
                    break; 
                }
            }

            return bestPair;
        }
    }
}

File: Python/Inventory_and_Order_Processing/budget_finder.py
from sorting import sort_products_by_price

def find_closest_pair_to_budget(products, budget):
    if len(products) < 2:
        print("Not enough products to form a pair.")
        return None
    
    sorted_products = sort_products_by_price(products)
    
    left = 0
    right = len(sorted_products) - 1
    
    best_pair = None
    best_difference = float("inf")
    
    while left < right:
        product_left = sorted_products[left]
        product_right = sorted_products[right]
        combined_price = product_left._price + product_right._price
        difference = abs(combined_price - budget)
        
        if difference < best_difference:
            best_difference = difference
            best_pair = (product_left, product_right, combined_price, difference)
            
        if combined_price < budget:
            left+=1
        elif combined_price > budget:
            right -= 1
        else:
            break
        
    return best_pair